wire-ios/Wire-iOS Tests/DeviceManagement/DeviceDetailsViewActionsHandlerTests.swift

```diff
@@ -0,0 +1,99 @@
+
+    override func setUp() {
+        super.setUp()
+        coreDataFixture = CoreDataFixture()
+        client = mockUserClient()
+        mockSession = UserSessionMock(mockUser: .createSelfUser(name: "Joe"))
+        emailCredentials = ZMEmailCredentials(email: "test@rad.com", password: "smalsdldl231S#")
+    }
+
+    func testGivenCertificateIsEmptyWhenDownladActionIsInvokedThenSaveFileManagerSaveFileIsNotCalled() {
+        let expectation = expectation(description: "Download file should be called")
+        expectation.isInverted = true
+        saveFileManager.saveValueIsCalled = { _, _, _ in
+            expectation.fulfill()
+        }
+        let deviceActionHandler = DeviceDetailsViewActionsHandler(
+            userClient: client,
+            userSession: mockSession,
+            credentials: emailCredentials,
+            e2eIdentityProvider: MockE2eIdentityProvider(),
+            mlsProvider: MockMLSProvider(isMLSEnbaled: false),
+            saveFileManager: saveFileManager
+        )
+        deviceActionHandler.downloadE2EIdentityCertificate(certificate: .mock())
+        wait(for: [expectation], timeout: 0.5)
+    }
+
+    func testGivenCertificateWhenDownladActionIsInvokedThenSaveFileManagerSaveFileIsCalled() {
+        let expectation = expectation(description: "Download file should be called")
+        saveFileManager.saveValueIsCalled = { _, _, _ in
+            expectation.fulfill()
+        }
+        let deviceActionHandler = DeviceDetailsViewActionsHandler(
+            userClient: client,
+            userSession: mockSession,
+            credentials: emailCredentials,
+            e2eIdentityProvider: MockE2eIdentityProvider(),
+            mlsProvider: MockMLSProvider(isMLSEnbaled: false),
+            saveFileManager: saveFileManager
+        )
+        deviceActionHandler.downloadE2EIdentityCertificate(certificate: .mock())
+        wait(for: [expectation], timeout: 0.5)
+    }
+
+    func testWhenFetchCertificateIsInvokedThenValidCertificateIsReturned() async {
+        let e2eIdentityProvider = MockE2eIdentityProvider()
+        let deviceActionHandler = DeviceDetailsViewActionsHandler(
+            userClient: client,
+            userSession: mockSession,
+            credentials: emailCredentials,
+            e2eIdentityProvider: e2eIdentityProvider,
+            mlsProvider: MockMLSProvider(isMLSEnbaled: false),
+            saveFileManager: MockSaveFileManager()
+        )
+        let fetchedCertificate = await deviceActionHandler.fetchCertificate()
+        XCTAssertNotNil(fetchedCertificate)
+        XCTAssertEqual(fetchedCertificate!.certificateDetails, e2eIdentityProvider.certificate.certificateDetails)
+    }
+}
+
+private extension E2eIdentityCertificate {
+
+    static func mock() -> E2eIdentityCertificate {
+        return .init(
+            certificateDetails: .random(length: 100),
+            expiryDate: .now,
+            certificateStatus: "valid",
```

Comment:
I would use a strong type instead of a string for the certificate status.

Reply:
That would be part of a different PR